pkg/scheduler/algorithm/generic.go

```diff
@@ -102,9 +112,84 @@ func (g *genericScheduler) Schedule(ctx context.Context, fwk framework.Framework
 	}, err
 }
 
+func estimateReplicas(ctx context.Context, fwk framework.Framework, sub *appsapi.Subscription, finv *appsapi.FeedInventory, clusters []*clusterapi.ManagedCluster) (framework.ClusterScoreList, error) {
+	availableList := make(framework.ClusterScoreList, len(clusters))
+	for i := range clusters {
+		availableList[i] = framework.ClusterScore{
+			NamespacedName: klog.KObj(clusters[i]).String(),
+		}
+	}
+
+	// if not dynamic dividing
+	if sub.Spec.SchedulingStrategy != appsapi.DividingSchedulingStrategyType {
+		return availableList, nil
+	}
+
+	// initialize
+	initialReplicas := make(framework.FeedReplicas, len(finv.Spec.Feeds))
+	for i := range finv.Spec.Feeds {
+		if finv.Spec.Feeds[i].DesiredReplicas != nil {
+			initialReplicas[i] = math.MaxInt32
+		}
+	}
+
+	// set boundary
+	for i := range clusters {
+		replicaCopy := make([]int32, len(initialReplicas))
+		copy(replicaCopy, initialReplicas)
+		availableList[i].MaxAvailableReplicas = replicaCopy
+	}
+
+	// reset boundary
+	defer func() {
+		for i := range availableList {
+			for j := range availableList[i].MaxAvailableReplicas {
+				if availableList[i].MaxAvailableReplicas[j] == math.MaxInt32 {
+					availableList[i].MaxAvailableReplicas[j] = *finv.Spec.Feeds[j].DesiredReplicas
+				}
+			}
+		}
+	}()
+
+	if !fwk.HasEstimatePlugins() {
+		return availableList, nil
+	}
+
+	// if not dynamic dividing
+	if sub.Spec.DividingScheduling == nil || sub.Spec.DividingScheduling.Type == appsapi.StaticReplicaDividingType {
+		return availableList, nil
+	}
+
+	// Run PreEstimate plugins.
+	preEstimateStatus := fwk.RunPreEstimatePlugins(ctx, sub, finv, clusters)
+	if !preEstimateStatus.IsSuccess() {
+		return nil, preEstimateStatus.AsError()
+	}
+
+	// Run the Estimate plugins.
+	estimateMap, estimateStatus := fwk.RunEstimatePlugins(ctx, sub, finv, clusters)
```

Comment:
Will EstimatePlugin return a value of 0, and will this cluster be filtered out in subsequent processes?

Reply:
No, `Filter` is finished when `Estimate` happens. We will not assign any replicas in this 0 available replicas in the following `Assign` step. However, if the target object is not a workload, i.e. whose `DesiredReplicas` pointer is nil, we still propagate it because non-workload object does not concern about estimation result.